pkg/coveragedb/coveragedb.go

```diff
@@ -627,3 +622,58 @@ func UniqCoverage(fullCov, partCov map[int]int64) map[int]int64 {
 	}
 	return res
 }
+
+func RegenerateSubsystems(ctx context.Context, ns string, sss []*subsystem.Subsystem,
+	client spannerclient.SpannerClient) (int, error) {
+	ssMatcher := subsystem.MakePathMatcher(sss)
+	ssCache := make(map[string][]string)
+	filePaths, err := getFilePaths(ctx, ns, client)
+	if err != nil {
+		return 0, err
+	}
+	var mutations []*spanner.Mutation
+	for _, filePath := range filePaths {
+		subsystems := getFileSubsystems(filePath, ssMatcher, ssCache)
+		mutations = append(mutations, fileSubsystemsMutation(ns, filePath, subsystems))
```

Comment:
Inside `fileSubsystemsMutation`, you create a InsertOrUpdate mutation. I am curious whether it only looks at the primary keys or at all the data? If it considers all fields and you try to update file's subsystem list, you could end up keeping the old row and inserting a new one each time.